Apply min-preludes to more tests (part 5)

danakj opened the pull request:
This drops file_test runtime from about 4s to 3.5s on my machine.

danakj committed:
min-preludes5

jonmeow commented on toolchain/check/testdata/deduce/binding_pattern.carbon:
```diff
@@ -2,9 +2,9 @@
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/convert.carbon
 // TODO: Add ranges and switch to "--dump-sem-ir-ranges=only".
 // EXTRA-ARGS: --dump-sem-ir-ranges=if-present
-// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/convert.carbon
```
TBH I prefer keeping the INCLUDE-FILE after (i.e. just to keep things sorted, lacking a better ordering), I just haven't been commenting because I expect most of the TODOs to be removed. But, it feels a bit odd to me to be touching a file just to move it?

danakj replied:
Hm, my thought here was that almost every test should eventually have `INCLUDE-FILE` in it, so it's best to come first, as this would aid copying tests to make new tests, and aid in discoverability. And it makes it much easier to quickly visually scan a file and see if it has `INCLUDE-FILE` in it yet, and makes tests without it stand out more clearly, which I considered as desirable. It is true it's not alphabetical. WDYT?

jonmeow replied:
We put AUTOUPDATE last, and it's in almost (but not all) every file, so I can't say that the "put more common things first" argument feels convincing to me. A different argument to make would be that it should be adjacent to AUTOUPDATE because then the boilerplate things are out of the way.

danakj replied:
Yeah, `AUTOUPDATE` has to be last cuz it gets `TIP`s placed below it. The copyright and INCLUDE-FILE end up being the header, and AUTOUPDATE being the footer, and exceptional stuff goes in between. We'll never get them alphabetical since `AUTOUPDATE` has to come last. Idk why `EXTRA-ARGS` going first feels wrong wrong to me. I guess I would like to be somewhat consistent either way, and alphabetical isn't going to work anyway.

jonmeow commented on toolchain/check/testdata/interface/compound_member_access.carbon:
```diff
@@ -2,9 +2,9 @@
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/facet_types.carbon
 // TODO: Add ranges and switch to "--dump-sem-ir-ranges=only".
 // EXTRA-ARGS: --dump-sem-ir-ranges=if-present
-// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/facet_types.carbon
```
As above

jonmeow commented on toolchain/check/testdata/interface/default_fn.carbon:
```diff
@@ -2,9 +2,9 @@
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/convert.carbon
 // TODO: Add ranges and switch to "--dump-sem-ir-ranges=only".
 // EXTRA-ARGS: --dump-sem-ir-ranges=if-present
-// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/destroy.carbon
```
As above

jonmeow commented on toolchain/check/testdata/interface/fail_assoc_const_not_constant.carbon:
```diff
@@ -2,7 +2,7 @@
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
-// EXTRA-ARGS: --no-prelude-import
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/convert.carbon
```
Why does this change? It doesn't result in changes to the test, is any prelude required?

danakj replied:
Done, there's no conversions in this file.

jonmeow commented on toolchain/check/testdata/interface/generic_method.carbon:
```diff
@@ -2,9 +2,9 @@
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/convert.carbon
 // TODO: Add ranges and switch to "--dump-sem-ir-ranges=only".
 // EXTRA-ARGS: --dump-sem-ir-ranges=if-present
-// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/destroy.carbon
```
Two things here:

1) I think the destroy prelude is smaller, so I'm not sure why this is changing
2) Relocating the line feels like a bikeshed, as above

danakj replied:
This test actually involves conversions, such as `{}` to `Y`, so I think keeping convert here seems better - even if destroy works for now.

jonmeow commented on toolchain/check/testdata/interface/name_poisoning.carbon:
```diff
@@ -2,7 +2,7 @@
 // Exceptions. See /LICENSE for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
-// EXTRA-ARGS: --no-prelude-import
```
Why does this change?

danakj replied:
Done, this file has no conversions

jonmeow reviewed (commented):
Why are you switching more minimal preludes (both no prelude and destroy.carbon) over to convert.carbon?

danakj commented:
> Why are you switching more minimal preludes (both no prelude and destroy.carbon) over to convert.carbon?

I am trying to standardize on `convert.carbon` being the common case when a test doesn't need anything in particular out of the prelude because:
- It is very tiny: 2 interfaces and an impl.
- It prevents bad errors in tests that do conversion implicitly or explicitly, which is a good default.

It's true that destroy will be smaller by 1 interface, given its current TODOs, but I don't know that that difference is too meaningful - debugging a test with 2 interfaces vs 1 is not a problematic difference in my experience and matches the voiced opinion of others that I recall when we discussed min-preludes. And the benefit to error diagnostics seems to make convert.carbon a better choice than destroy.carbon, unless you need destruction in your test.

I would actually argue that maybe destroy.carbon should include as.carbon to also provide better error diagnostics.

I am planning to document the use of convert.carbon when nothing from the prelude is needed as recommended policy in the readme.

Then for no-prelude, my sense is that there are two kinds of tests written as no-prelude right now:
- Tests that didn't need anything from the prelude and were written as no-prelude to be faster/more minimal. Changing them to convert.carbon achieves the same goals and keeps them nice for debugging, while improving error diagnostics when causing conversion errors in new or existing tests in the file, and matches a good default practice (IMHO).
- Tests that are testing a no-prelude environment, because they are reproducing things from the prelude or testing a bare-metal environment. These should stay no-prelude. There are some tests in this category that end up relying on some other unrelated part of the prelude like i32 FBOFW, and probably would be better written without that, but that's an aside.

WDYT?

jonmeow commented:
> > Why are you switching more minimal preludes (both no prelude and destroy.carbon) over to convert.carbon?
> 
> I am trying to standardize on `convert.carbon` being the common case when a test doesn't need anything in particular out of the prelude because:
> 
> * It is very tiny: 2 interfaces and an impl.
> * It prevents bad errors in tests that do conversion implicitly or explicitly, which is a good default.
> 
> It's true that destroy will be smaller by 1 interface, given its current TODOs, but I don't know that that difference is too meaningful - debugging a test with 2 interfaces vs 1 is not a problematic difference in my experience and matches the voiced opinion of others that I recall when we discussed min-preludes. And the benefit to error diagnostics seems to make convert.carbon a better choice than destroy.carbon, unless you need destruction in your test.
> 
> I would actually argue that maybe destroy.carbon should include as.carbon to also provide better error diagnostics.

Destruction probably shouldn't be casting-based. What error diagnostics are you expecting to change? Can you give examples?

(BTW, "unless you need destruction in your test" -- this is the precise issue being addressed)

> I am planning to document the use of convert.carbon when nothing from the prelude is needed as recommended policy in the readme.
> 
> Then for no-prelude, my sense is that there are two kinds of tests written as no-prelude right now:
> 
> * Tests that didn't need anything from the prelude and were written as no-prelude to be faster/more minimal. Changing them to convert.carbon achieves the same goals and keeps them nice for debugging, while improving error diagnostics when causing conversion errors in new or existing tests in the file, and matches a good default practice (IMHO).
> * Tests that are testing a no-prelude environment, because they are reproducing things from the prelude or testing a bare-metal environment. These should stay no-prelude. There are some tests in this category that end up relying on some other unrelated part of the prelude like i32 FBOFW, and probably would be better written without that, but that's an aside.
> 
> WDYT?

no-prelude is offering something else: fewer dependencies. That way, if something fails, we can look at using that as a way to break apart failures. Conversely, if a dependency is *accidentally* added, we have more warning flags.

So I think no-prelude is actually very important.

danakj committed:
leave-destroy

danakj committed:
no-preludes

danakj commented:
> > > Why are you switching more minimal preludes (both no prelude and destroy.carbon) over to convert.carbon?
> > 
> > 
> > I am trying to standardize on `convert.carbon` being the common case when a test doesn't need anything in particular out of the prelude because:
> > 
> > * It is very tiny: 2 interfaces and an impl.
> > * It prevents bad errors in tests that do conversion implicitly or explicitly, which is a good default.
> > 
> > It's true that destroy will be smaller by 1 interface, given its current TODOs, but I don't know that that difference is too meaningful - debugging a test with 2 interfaces vs 1 is not a problematic difference in my experience and matches the voiced opinion of others that I recall when we discussed min-preludes. And the benefit to error diagnostics seems to make convert.carbon a better choice than destroy.carbon, unless you need destruction in your test.
> > I would actually argue that maybe destroy.carbon should include as.carbon to also provide better error diagnostics.
> 
> Destruction probably shouldn't be casting-based. What error diagnostics are you expecting to change? Can you give examples?
> 
> (BTW, "unless you need destruction in your test" -- this is the precise issue being addressed)
> 
> > I am planning to document the use of convert.carbon when nothing from the prelude is needed as recommended policy in the readme.
> > Then for no-prelude, my sense is that there are two kinds of tests written as no-prelude right now:
> > 
> > * Tests that didn't need anything from the prelude and were written as no-prelude to be faster/more minimal. Changing them to convert.carbon achieves the same goals and keeps them nice for debugging, while improving error diagnostics when causing conversion errors in new or existing tests in the file, and matches a good default practice (IMHO).
> > * Tests that are testing a no-prelude environment, because they are reproducing things from the prelude or testing a bare-metal environment. These should stay no-prelude. There are some tests in this category that end up relying on some other unrelated part of the prelude like i32 FBOFW, and probably would be better written without that, but that's an aside.
> > 
> > WDYT?
> 
> no-prelude is offering something else: fewer dependencies. That way, if something fails, we can look at using that as a way to break apart failures. Conversely, if a dependency is _accidentally_ added, we have more warning flags.
> 
> So I think no-prelude is actually very important.

I have replaced no-prelude, as those tests did not have conversions. And replaced destroy.carbon in one test which had no conversions. The other two do convert, so I left them as convert.carbon.
